### ISSUE TITLE
Use kotlin compiler plugin and reflection to generate code [Moved from old repo]

### ISSUE THREAD

**Issue by chippmann** [utopia-rise/godot-kotlin#2]
**Describe the problem or limitation you are having in your project:**
Currently the `api-classes-generator` and the `entry-generator` use a errorprone and hard to understand code generation using stringBuilders.
The code would be much cleaner and easier to maintain if we use reflection to generate the api-wrapper and annotations for the entry-classes generation.
Also it would make the project easier to use as a user doesn't have to write config json's anymore.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
One no longer needs to configure the errorprone json files and maitanance is much simpler.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
None to provide.

**Describe implementation detail for your proposal (in code), if possible:**
Use the compiler plugin created in issue #13 for the config generation.
~~Maybe use [kotlinPoet](https://github.com/square/kotlinpoet) for the api classes generation~~ -> done in #14 

**If this enhancement will not be used often, can it be worked around with a few lines of code?:**
Yes with the current system. But it's not user friendly at all.

**Is there a reason why this should be in this project and not individually solved?:**
Yes as the code generation is the main part of the project.

**Comment by chippmann** [comment]
i wrote in old repo:

Summarizing some discussions regarding this issue on discord:
Currently we have an annotation processor implemeted. Annotation processing with kotlin multiplatform and kapt only works if one has a `jvm` target. Our idea was to create a fake jvm target in the `godot-gradle-plugin` so the user can write his code in `commonMain` and we then can process the annotations through the fake `jvm` target to generate the registration code.
This works but leads to the fact that the user has to write a second native module if he want's to use some native libraries.
This and the fact that the user always writes native code anyways (as godot needs a native binary at the end), we decided that the best approach would be to implement a kotlin compiler plugin in which we do the annotation processing and thus code generation.
This would lead to complete freedom for the user as he can decide in which module he want's to write his code in.

Currently I'm now implementing this annotation processor.

Edit to the issue: moving annotation processor and thus compiler plugin implementation in a separate issue (#13).

**Comment by chippmann** [comment]
i wrote in old repo:

Basic entry generation is done. I still need to do the following:  
- [x] Code cleanup
- [x] Sanity checks
- [x] Remove old entry generation completely
- [ ] ~~Write unit tests~~ <- will be done at a later time

@piiertho I noted that even though the RPC stuff was documented, it was actually never properly implemented. There is a stale branch which did the basics with a new example: https://github.com/utopia-rise/kotlin-godot-wrapper/tree/withRPC
If you are ok with it, I'll implement all networking related registration logic in this issue as well.

**Comment by chippmann** [comment]
Summation of discussions on Discord:
Everything is working now. But i have the problem that the MpApt Annotation Processor is too late in the compilation process to add the generated entry file to the sources to be compiled. This results in the need to build the project twice so the generated entry file is included in the final binary.

We discussed the problem on discord and i maybe found a solution (message from discord):  

---
I tried what i talked about yesterday and it seems to work.
Basically i just made a test script. I'm not yet 100% sure if it would work but before i continue i want to hear your opinion.
Basically what i've done is extracted part of the compiler frontend into a script so we can use it directly like we want: https://github.com/utopia-rise/godot-kotlin/blob/feature/rework_entry_code_generation/tools/entry-generator/src/main/kotlin/org/godotengine/kotlin/entrygenerator/parser/SourceCodeParser.kt
Important in this script is the parse function.

When running it, we get almost the same output as in the annotation processor and can process the information.

My idea:
Make a gradle task that runs before build. In such a gradle task we know which source sets are defined and can access all the source files in it.
The path's to those sourcefiles we pass to the extraced compiler frontend in that script and generate the entry file from it.
then the normal build runs and correctly detects our entry file.

Note: the script is just a test script and need's cleaning up but it does the job.

Why i'm not yet sure if this would work:
- I didn't test every class in the samples yet

Drawback of this solution:
On compiler API changes i also need to change this script, but i think that's not that big of a deal as we only need to change it if we wan't to and is not necessary to bumpt the same way we bump the kotlin version on the other parts of the project.

otherwise i'm pretty sure it would work.  

---
I'll continue with above described solution for now to see if it works.

**Comment by chippmann** [comment]
Some updates (already shared through discord):
The above mentioned solution with the extraction of the compiler frontend works for the most part. But an important part does not. At the moment I cannot read annotation arguments as long as I don't add the source file's of the annotations to the parsing process (adding them to the classpath of the parsing source files does not work for some reason).
@piiertho had the idea of adding the annotations sources through the source jar manually. I'm trying to implement this now.
I this does not work. We probably have to rely on the expect/actual implementation in #20 and use a default jvm target to be able to use kapt annotation processing. But we only take this route as a last resort.

**Comment by chippmann** [comment]
Some updates:
I got the parsing working with the idea from @piiertho. But it takes upfront 30s to just generate the entry file. So it's not feasable.
I now switched back to the previous attempt of using the compiler plugin.
By using a fake target which we define at built time in the gradle plugin, we compile the fake target first which generates the entry file and then run the actual compilation for the users targets to build everything.
As the compilation step is acually the shortest step atm (the linking step is the one that takes time) the time overhead seems negliable.
I implemented that now and my buildtime incresed from 1min 15s to 1min 30s.

Another good thing about this approach would be that it allows for easy switching to normal annotation processing once #20 and #10 is done.

I'll finish this now and we should have a beta to test up at around wednesday next week.
